src/frontend/src/optimizer/plan_expr_rewriter/const_eval_rewriter.rs

```diff
@@ -0,0 +1,50 @@
+// Copyright 2023 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_common::error::RwError;
+
+use crate::expr::{Expr, ExprImpl, ExprRewriter, Literal};
+
+pub(crate) struct ConstEvalRewriter {
+    pub(crate) error: Option<RwError>,
+}
+impl ExprRewriter for ConstEvalRewriter {
+    fn rewrite_expr(&mut self, expr: ExprImpl) -> ExprImpl {
+        if self.error.is_some() {
+            return expr;
+        }
+        if expr.is_const() {
+            let data_type = expr.return_type();
+            match expr.eval_row_const() {
```

Comment:
I'm still a little bit worried about the implementation of `eval_row_const`. But maybe it's not a big deal and the interface doesn't have any problem 🥵

Reply:
Yes, there is overhead to it... But at least we evaluate it just once, whereas if we propagate non-const-folded exprs to e.g. stream jobs, then we will incur linear overhead.